src/uvwasi.c

```diff
@@ -2593,30 +2699,187 @@ uvwasi_errno_t uvwasi_sock_send(uvwasi_t* uvwasi,
                                 uvwasi_size_t si_data_len,
                                 uvwasi_siflags_t si_flags,
                                 uvwasi_size_t* so_datalen) {
-  /* TODO(cjihrig): Waiting to implement, pending
-                    https://github.com/WebAssembly/WASI/issues/4 */
-  UVWASI_DEBUG("uvwasi_sock_send(uvwasi=%p, unimplemented)\n", uvwasi);
-  return UVWASI_ENOTSUP;
-}
 
+  struct uvwasi_fd_wrap_t *wrap;
+  uvwasi_errno_t err = 0;
+  uv_buf_t* bufs;
+  int r = 0;
+
+  if (uvwasi == NULL || si_data == NULL || so_datalen == NULL)
+    return UVWASI_EINVAL;
+
+  err = uvwasi_fd_table_get(uvwasi->fds,
+                            sock,
+                            &wrap,
+                            UVWASI__RIGHTS_SOCKET_BASE,
+                            0);
+  if (err != UVWASI_ESUCCESS)
+    return err;
 
+  err = uvwasi__setup_ciovs(uvwasi, &bufs, si_data, si_data_len);
+  if (err != UVWASI_ESUCCESS) {
+    uv_mutex_unlock(&wrap->mutex);
+    return err;
+  }
+
+  r = uv_try_write((uv_stream_t*) wrap->sock, bufs, si_data_len);
+  uvwasi__free(uvwasi, bufs);
+  uv_mutex_unlock(&wrap->mutex);
+  if (r < 0)
+    return uvwasi__translate_uv_error(r);
+
+  *so_datalen = (uvwasi_size_t) r;
+  return UVWASI_ESUCCESS;
+}
+
+typedef struct shutdown_data_s {
```

Comment:
Can we move some of these structs and helper functions to a separate file.

Reply:
Done